### ISSUE TITLE
aten::local_scalar_dense

### ISSUE THREAD

**Issue by MaxGhi8** [pytorch/xla#6193]
## ❓ Questions and Help
Hello, I encountered an issue with the function aten:local_scalar_dense. In particular using the pytorch profiler I see that the command `coefficients[0:-1:coefficients.size(0)] *= torch.tensor(2)` increase a lot the usage of aten::local_scalr_dense, and that cause an augmentation of 1 seconds per epochs ... It is annoying because this is one of the simplest operation of the code and it takes approximately 1/8 of the total times of each epoch ... Anyone knows how to fix it?

p.s. the code `coefficients[0:-1:coefficients.size(0)] *= torch.tensor(2)` correspond to `coefficients[0] *= 2` and `coefficients[-1] *= 2`.

**Comment by JackCaoG** [comment]
I tried a simple repo
```
>>> coefficients = torch.randn(5, 5, device='xla:0')
>>> import torch_xla.debug.metrics as met
>>> print(met.short_metrics_report())
Metric: TransferToServerTime
  TotalSamples: 1
  Accumulator: 029ms430.817us
  Percentiles: 1%=029ms430.817us; 5%=029ms430.817us; 10%=029ms430.817us; 20%=029ms430.817us; 50%=029ms430.817us; 80%=029ms430.817us; 90%=029ms430.817us; 95%=029ms430.817us; 99%=029ms430.817us
>>> coefficients[0:-1:coefficients.size(0)] *= torch.tensor(2)
>>> print(met.short_metrics_report())
Metric: TransferToServerTime
  TotalSamples: 2
  Accumulator: 029ms497.947us
  ValueRate: 461.677us / second
  Rate: 0.0313023 / second
  Percentiles: 1%=067.130us; 5%=067.130us; 10%=067.130us; 20%=067.130us; 50%=029ms430.817us; 80%=029ms430.817us; 90%=029ms430.817us; 95%=029ms430.817us; 99%=029ms430.817us
```

I didn't see any `local_scalar_dense` in our metrics. `local_scalar_dense` usually means that pytorch as XLA to move certain tensor back to CPU().

For example if I do 
```
>>> print(coefficients[0][0].item())
```
I can see
```
>>> print(met.short_metrics_report())
.....
......
Counter: aten::_local_scalar_dense
  Value: 1
```

Can you provide a small repo?

**Comment by MaxGhi8** [comment]
Meanwhile, thank you very much for your reply. I am implementing the Fourier Neural Operator (FNO), only I would like to replace the fourier transform (fft.rfft2 in pytorch) with the Chebyshev transform (which I have tried to implement from ftt.fft). I noticed that the execution times for each epoch of my implementation are about double those of the original FNO, to investigate where the problem was I used pytorch's profiler and obtained the following two pictures.
![Immagine 2023-12-19 105048](https://github.com/pytorch/xla/assets/108077874/5e7387d5-5cd1-4df9-8d47-fbe5b2509ab8)
![Immagine 2023-12-19 104842](https://github.com/pytorch/xla/assets/108077874/5f62e985-6edb-4a1f-901f-74ec6d24167e)
The first one is the original FNO and the second one is my implementation with the Chebyshev transform. I implemented the Chebyshev transform and antitrasform (that is the main difference between the two codes) as:
`def values_to_coefficients(values):
    # input array has shape `(n1, n2, ..., c)`, where c stands for channels
    # transform values of the function on the Chebyshev grid to coefficients of Chebyshev series
    D = values.shape
    for N in D[:-1]:
        values = torch.cat((values, values[1:-1].flip(0)), dim = 0)
        values = torch.real(torch.fft.rfft(values, dim=0))/(N-1)
        values[0:-1:values.size(0)] /= torch.tensor(2, device = mydevice) # values[0] /= 2 # values[-1] /= 2
        values = values * ((-1)**torch.arange(N, dtype = torch.int)).reshape([-1]+[1]*(len(D)-1))
        values = torch.permute(values, tuple([i for i in range(1, len(D)-1)]) + tuple((0, len(D)-1)))

    return values

batched_values_to_coefficients = torch.vmap(values_to_coefficients)

def coefficients_to_values(coefficients):
    # x = coefficients_to_values(values_to_coefficients(x)) up to round-off error
    D = coefficients.shape
    for N in D[:-1]:
        coefficients = coefficients * ((-1)**torch.arange(N, dtype=torch.int)).reshape([-1]+[1]*(len(D)-1))
        coefficients[0:-1:coefficients.size(0)] *= torch.tensor(2, device = mydevice) # coefficients[0] *= 2 # coefficients[-1] *= 2
        coefficients = torch.cat(((N-1)*coefficients, (N-1)*coefficients[1:-1].flip(0)), dim = 0)
        coefficients = torch.real(torch.fft.rfft(coefficients, dim=0))/(N-1)/2
        coefficients = torch.permute(coefficients, tuple([i for i in range(1, len(D)-1)]) + tuple((0, len(D)-1)))
    
    return coefficients

batched_coefficients_to_values = torch.vmap(coefficients_to_values)`

Not being an expert, I tried commenting where I do multiplication and division by two and get that the function `aten::local_scalar_dense` is reduced by a lot, which is why I asked the previous question.
![Immagine 2023-12-19 104408](https://github.com/pytorch/xla/assets/108077874/f16fabec-0746-4ba9-9de6-922fd70c7c8e)

In general, any suggestions or references for learning something about this problem are very welcome and I thank you again very much.

**Comment by JackCaoG** [comment]
@Liyang90 Do you have any insight why
```
coefficients[0:-1:coefficients.size(0)] *= torch.tensor(2, device = mydevice)
```
will cause `local_scalar_dense `?

@MaxGhi8 you can take a look at https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#compilation--execution-analysis If you are using nightly(or 2.2 which will be release in Jan), enabling `PT_XLA_DEBUG=1` will let you know when compilatiion/execution happens. `local_scalar_dense` will usually trigger a graph execution since it try to move a value from device to the host and the value might requires computation(since `mark_step` has not been called yet). More docs can be found in https://github.com/pytorch/xla?tab=readme-ov-file#additional-reads and https://github.com/pytorch/xla/tree/master/docs

**Comment by Liyang90** [comment]
I don't see an obvious red flag that could explain the `local_scalar_dense`. Probably going through the debug steps to find the offending ops is needed.

**Comment by MaxGhi8** [comment]
Ok, thank you very much for the suggestions. I will try and let you know if I find anything interesting. Thank you!
